### PR TITLE
Issue/dev 2773/api token management improvement

### PR DESCRIPTION
**Pull request by edoardoc**
https://tracker.rport.io/issue/DEV-2773/api-token-management-improvement

### CODE REVIEW

File: db/migration/api_token/sql/001_init.up.sql [hunk 1]
```diff
@@ -1,9 +1,13 @@
 CREATE TABLE api_token (
```
Comment (thorstenkramm) [inline db/migration/api_token/sql/001_init.up.sql:1]: If the endpoint is plural, the table must use plural too. `api_tokens`.

File: db/migration/api_token/sql/001_init.up.sql [hunk 2]
```diff
@@ -1,9 +1,13 @@
...
     created_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP, 
     expires_at DATETIME,
     scope TEXT,
     token TEXT NOT NULL,
     PRIMARY KEY (username, prefix)
-) WITHOUT ROWID;
\ No newline at end of file
+) WITHOUT ROWID;
+
+CREATE UNIQUE INDEX api_token_unique_name
```
Comment (thorstenkramm) [inline db/migration/api_token/sql/001_init.up.sql:12]: see above

File: api-doc/openapi/paths/me_token.yaml [hunk 3]
```diff
@@ -53,6 +53,9 @@ put:
             prefix:
```
Comment (thorstenkramm) [inline api-doc/openapi/paths/me_token.yaml:53]: The API shall return just the token, including the prefix. The prefix is useless and irrelevant for the user.

**Review by thorstenkramm — CHANGES_REQUESTED** [review]

File: db/migration/api_token/sql/001_init.up.sql [hunk 4]
```diff
@@ -1,9 +1,13 @@
 CREATE TABLE api_token (
     username TEXT NOT NULL CHECK (username != ''),
     prefix TEXT NOT NULL CHECK (prefix != ''),
+    name TEXT NOT NULL,
```
Comment (domenipavec) [inline db/migration/api_token/sql/001_init.up.sql:4]: This migration won't be run for existing databases. You should create a new migration that adds a column.

File: server/api_handler_login_test.go [hunk 5]
```diff
@@ -174,7 +210,7 @@ func TestAPITokenOps(t *testing.T) {
 
 			// then
 			require.Equal(tc.wantStatusCode, w.Code)
-			if tc.wantErrTitle == "" {
+			if strings.HasPrefix(fmt.Sprintf("%d", tc.wantStatusCode), "2") { // any 2** status code is a success test case
```
Comment (domenipavec) [inline server/api_handler_login_test.go:213]: This is a bit weird way to test this, you could just do `tc.wantStatusCode/100 == 2`.

File: server/api_handler_login_test.go [hunk 6]
```diff
@@ -183,16 +219,25 @@ func TestAPITokenOps(t *testing.T) {
...
+
+				// decode response
+				tResponse := api.NewErrAPIPayloadFromMessage("", "", "")
+				err = json.Unmarshal(w.Body.Bytes(), &tResponse)
 				require.NoError(err)
-				require.Equal(string(wantRespBytes), w.Body.String())
+
+				// compare items of struct error, one by one,
+				if tc.wantErrTitle != "" {
+					require.Equal(tc.wantErrTitle, tResponse.Errors[0].Title)
```
Comment (domenipavec) [inline server/api_handler_login_test.go:230]: You should use `assert.Equal` so the test continues if this is not equal and we can see all failures.
`require` should be used in cases when it makes no sense to continue, like above if json unmarshal fails with errors it makes no sense to do these comparisons.

**Review by domenipavec — CHANGES_REQUESTED** [review]

File: db/migration/api_token/sql/001_init.up.sql (hunk 1 above)
Comment (edoardoc) [reply db/migration/api_token/sql/001_init.up.sql:1]: Entities should be named in a singular way - especially database ones and regardless whatever bs one does on the next layer.
I could speak about this for a while, though checking the codebase I understand there is the same mistake in all other tables so I will just follow the convention here.

File: server/api_handler_login_test.go (hunk 5 above)
Comment (edoardoc) [reply server/api_handler_login_test.go:213]: oups :)

File: server/api_handler_me.go [hunk 7]
```diff
@@ -277,11 +285,21 @@ func (al *APIListener) handlePostToken(w http.ResponseWriter, req *http.Request)
...
+	}
+
 	if r.Scope == authorization.APITokenClientsAuth && !user.IsAdmin() {
 		al.jsonErrorResponseWithTitle(w, http.StatusBadRequest, "current user should belong to Administrators group to create a token with this scope")
 		return
 	}
 
+	createdAt := ptr.Time(time.Now().Truncate(time.Second).UTC())
+	if r.ExpiresAt == nil {
+		r.ExpiresAt = ptr.Time(createdAt.AddDate(1, 0, 0))
```
Comment (domenipavec) [inline server/api_handler_me.go:300]: Is this a requirement to not have tokens without expiration date?

Can you add a comment so it's clear what is getting added? Like `AddDate(1 /* year */, 0, 0)`

Comment (edoardoc) [reply server/api_handler_me.go:300]: yes it is a requirement, default for expiry date is 1 year from creation date (if not provided)
I will add a comment

**Pull request merged**